src/os/bluestore/BlueStore.cc

```diff
@@ -7897,110 +7926,95 @@ int BlueStore::_fsck_check_extents(
   return errors;
 }
 
-void BlueStore::_fsck_check_pool_statfs(
-  BlueStore::per_pool_statfs& expected_pool_statfs,
+void BlueStore::_fsck_check_statfs(
+  const store_statfs_t& expected_statfs,
+  const per_pool_statfs& expected_pool_statfs,
   int64_t& errors,
   int64_t& warnings,
   BlueStoreRepairer* repairer)
 {
-  auto it = db->get_iterator(PREFIX_STAT, KeyValueDB::ITERATOR_NOCACHE);
-  if (it) {
-    for (it->lower_bound(string()); it->valid(); it->next()) {
-      string key = it->key();
-      if (key == BLUESTORE_GLOBAL_STATFS_KEY) {
-        if (repairer) {
-	  ++errors;
-	  repairer->remove_key(db, PREFIX_STAT, BLUESTORE_GLOBAL_STATFS_KEY);
-	  derr << "fsck error: " << "legacy statfs record found, removing"
-	       << dendl;
-	}
-	continue;
-      }
-      uint64_t pool_id;
-      if (get_key_pool_stat(key, &pool_id) < 0) {
-	derr << "fsck error: bad key " << key
-	     << "in statfs namespece" << dendl;
-	if (repairer) {
-	  repairer->remove_key(db, PREFIX_STAT, key);
-	}
-	++errors;
-	continue;
-      }
-
-      volatile_statfs vstatfs;
-      bufferlist bl = it->value();
-      auto blp = bl.cbegin();
-      try {
-	vstatfs.decode(blp);
-      } catch (ceph::buffer::error& e) {
-        derr << "fsck error: failed to decode Pool StatFS record"
-	     << pretty_binary_string(key) << dendl;
+  string key;
```

Comment:
I feel like we can make check/repair statfs much simpler.
We could create a map of {key->stat} from the stats we just scanned.
This should be ideally reflected by state of DB.
If DB does not have something - we complain and add (in repair mode).
If DB has different value - we complain and fix (in repair mode).
If DB has extra key - we complain and delete (in repair mode).

Reply:
The tricky thing is that now we might have statfs persisted at DB in two different ways depending on the previous shutdown: graceful or not. This implementation checks in-memory stats (loaded either from relevant DB records or via recovery) vs. stats loaded by fsck.
For your suggestion we'll need to perform DB update immediately after the recovery and then de-serialize it back for fsck (and then finally update DB once again on a shutdown). Which doesn't make much sense to me...